src/Lib.P2P/Discovery/Mdns.cs

```diff
@@ -0,0 +1,160 @@
+*/
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Common.Logging;
+using Makaretu.Dns;
+using MultiFormats;
+
+namespace Lib.P2P.Discovery
+{
+    /// <summary>
+    ///   Base class to discover peers using Multicast DNS.
+    /// </summary>
+    public abstract class Mdns : IPeerDiscovery
+    {
+        private static ILog log = LogManager.GetLogger(typeof(Mdns));
+
+        /// <inheritdoc />
+        public event EventHandler<Peer> PeerDiscovered;
+
+        /// <summary>
+        ///  The local peer.
+        /// </summary>
+        public Peer LocalPeer { get; set; }
+
+        /// <summary>
+        ///   The Muticast Domain Name Service to use.
+        /// </summary>
+        public MulticastService MulticastService { get; set; }
+
+        /// <summary>
+        ///   The service name for our peers.
+        /// </summary>
+        /// <value>
+        ///   Defaults to "ipfs".
+        /// </value>
+        public string ServiceName { get; set; } = "ipfs";
+
+        /// <summary>
+        ///   Determines if the local peer responds to a query.
+        /// </summary>
+        /// <value>
+        ///   <b>true</b> to answer queries.  Defaults to <b>true</b>.
+        /// </value>
+        public bool Broadcast { get; set; } = true;
+
+        /// <inheritdoc />
+        public Task StartAsync()
+        {
+            MulticastService.NetworkInterfaceDiscovered += (s, e) =>
+            {
+                try
+                {
+                    var profile = BuildProfile();
+                    var discovery = new ServiceDiscovery(MulticastService);
+                    OnServiceDiscovery(discovery);
+                    discovery.ServiceInstanceDiscovered += OnServiceInstanceDiscovered;
+
+                    if (Broadcast && profile != null)
+                    {
+                        log.Debug($"Advertising {profile.FullyQualifiedName}");
+                        discovery.Advertise(profile);
+                    }
+
+                    // Ask all peers to broadcast discovery info.
+                    discovery.QueryServiceInstances(ServiceName);
+                }
+                catch (Exception ex)
+                {
+                    log.Debug("Failed to send query", ex);
+
+                    // eat it
+                }
+            };
+
+            return Task.CompletedTask;
+        }
+
+        /// <inheritdoc />
+        public Task StopAsync()
+        {
+            PeerDiscovered = null;
+            return Task.CompletedTask;
+        }
+
+        private void OnServiceInstanceDiscovered(object sender, ServiceInstanceDiscoveryEventArgs e)
+        {
+            try
+            {
+                var msg = e.Message;
+
+                // Is it our service?
+                var qsn = new DomainName(ServiceName + ".local");
+                if (!e.ServiceInstanceName.BelongsTo(qsn))
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)